File: dart_cli/lib/KeysLookup.dart

import "structs/KeysTable.dart";

const _globalTableEnd = 0x8000;
const _localTableOffset = 0x8001;

class KeysLookup {
  final KeysTable globalKeysTable;
  KeysTable? localKeysTable;

  KeysLookup(this.globalKeysTable, [this.localKeysTable]);

  String lookupIndex(int index) {
    if (index <= _globalTableEnd) {
      return globalKeysTable.lookupIndex(index);
    } else {
      if (localKeysTable == null)
        throw Exception("Local keys table not found");
      return localKeysTable!.lookupIndex(index - _localTableOffset);
    }
  }

  int getKeyIndex(String key) {
    var index = globalKeysTable.getKeyIndex(key);
    if (index != null)
      return index;
    localKeysTable ??= KeysTable();
    index = localKeysTable!.getKeyIndex(key, allowCreate: true);
    if (index == null)
      throw Exception("Could not create key in local keys table");
    return index + _localTableOffset;
  }

  void visitKey(String key) {
    getKeyIndex(key);
  }
}

File: python_lib/KeysLookup.py
from structs.KeysTable import KeysTable

_globalTableEnd = 0x8000
_localTableOffset = 0x8001

class KeysLookup:
	globalKeysTable: KeysTable
	localKeysTable: KeysTable|None

	def __init__(self, globalKeysTable: KeysTable, localKeysTable: KeysTable|None = None):
		self.globalKeysTable = globalKeysTable
		self.localKeysTable = localKeysTable

	def lookupIndex(self, index: int) -> str:
		if index <= _globalTableEnd:
			return self.globalKeysTable.lookupIndex(index)
		else:
			if self.localKeysTable is None:
				raise Exception("Local keys table not found")
			return self.localKeysTable.lookupIndex(index - _localTableOffset)
		
	def getKeyIndex(self, key: str) -> int:
		index = self.globalKeysTable.getKeyIndex(key)
		if index is not None:
			return index
		if self.localKeysTable is None:
			self.localKeysTable = KeysTable()
		index = self.localKeysTable.getKeyIndex(key, allowCreate=True)
		if index is None:
			raise Exception("Could not create key in local keys table")
		return index + _localTableOffset
	
	def visitKey(self, key: str):
		self.getKeyIndex(key)
